File: js/action/SFActionInterval.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  SFActionInterval.js
//  SpriteForest-HTML5
//
//  Created by moKy on Mar. 4, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	
	var prop = {
		ctor: sf.FiniteTimeAction.prototype.ctor,
		
		// properties
	};
	
	// extends
	sf.ActionInterval = sf.FiniteTimeAction.extend(prop);
	
}(SpriteForest);

//===================================================================== Sequence

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function Sequence() {
		sf.ActionInterval.prototype.ctor.call(this);
		this._actions = null;
	}
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function init() {
		if (!sf.ActionInterval.prototype.init.call(this)) {
			return false;
		}
		this._actions = null;
		return true;
	}
	function initWithDictionary(dict) {
		if (!sf.ActionInterval.prototype.initWithDictionary(dict)) {
			return false;
		}
		var actions = sf.Action.subActionsFromDictionary(dict);
		if (!actions) {
			cn.error(dict);
			return false;
		}
		// remove invalid actions
		for (var i = actions.length - 1; i >= 0; --i) {
			if (actions[i] instanceof cc.FiniteTimeAction) {
			} else {
				actions.splice(i, 1);
			}
		}
		
		if (actions.length < 1) {
			cn.error(dict);
			return false;
		}
		
		this._actions = actions;
		return true;
	}
	
	function ccAction() {
		if (!this._actions) {
			cn.error(this._actions);
			return null;
		}
		return cc.Sequence.create(this._actions);
	}
	
	var prop = {
		ctor: Sequence,
		
		init: init,
		initWithDictionary: initWithDictionary,
		
		ccAction: ccAction,
		
		// properties
		_actions: null,
	};
	
	// extends
	sf.Sequence = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

//======================================================================= Repeat

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function Repeat() {
		sf.ActionInterval.prototype.ctor.call(this);
		this._action = null;
		this._times = 0;
	}
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function init() {
		if (!sf.ActionInterval.prototype.init.call(this)) {
			return false;
		}
		this._action = null;
		this._times = 0;
		return true;
	}
	function initWithDictionary(dict) {
		if (!sf.ActionInterval.prototype.initWithDictionary.call(this, dict)) {
			return false;
		}
		var action = sf.Action.subActionFromDictionary(dict);
		if (action instanceof cc.FiniteTimeAction) {
			this._action = action;
		}
		this._times = dict['Times'];
		return true;
	}
	
	function ccAction() {
		if (!this._action || !this._times) {
			cn.error(this._action, this._times);
			return null;
		}
		return cc.Repeat.create(this._action, this._times);
	}
	
	var prop = {
		ctor: Repeat,
		
		init: init,
		initWithDictionary: initWithDictionary,
		
		ccAction: ccAction,
		
		// properties
		_action: null,
		_times: 0,
	};
	
	// extends
	sf.Repeat = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

//======================================================================== Spawn

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function Spawn() {
		sf.ActionInterval.prototype.ctor.call(this);
		this._actions = null;
	}
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function init() {
		if (!sf.ActionInterval.prototype.init.call(this)) {
			return false;
		}
		this._actions = null;
		return true;
	}
	function initWithDictionary(dict) {
		if (!sf.ActionInterval.prototype.initWithDictionary.call(this, dict)) {
			return false;
		}
		var actions = sf.Action.subActionsFromDictionary(dict);
		if (!actions) {
			cn.error(dict);
			return false;
		}
		// remove invalid actions
		for (var i = actions.length - 1; i >= 0; --i) {
			if (actions[i] instanceof cc.FiniteTimeAction) {
			} else {
				actions.splice(i, 1);
			}
		}
		
		if (actions.length < 1) {
			cn.error(dict);
			return false;
		}
		
		this._actions = actions;
		return true;
	}
	
	function ccAction() {
		if (!this._actions) {
			cn.error(this._actions);
			return null;
		}
		return cc.Spawn.create(this._actions);
	}
	
	var prop = {
		ctor: Spawn,
		
		init: init,
		initWithDictionary: initWithDictionary,
		
		ccAction: ccAction,
		
		// properties
		_actions: null,
	};
	
	// extends
	sf.Spawn = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

//==================================================================== DelayTime

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function ccAction() {
		return cc.DelayTime.create(this._action);
	}
	
	var prop = {
		ctor: sf.ActionInterval.prototype.ctor,
		
		ccAction: ccAction,
		
		// properties
	};
	
	// extends
	sf.DelayTime = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

//================================================================== ReverseTime

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function ReverseTime() {
		sf.ActionInterval.prototype.ctor.call(this);
		this._action = null;
	}
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function init() {
		if (!sf.ActionInterval.prototype.init.call(this)) {
			return false;
		}
		this._action = null;
		return true;
	}
	function initWithDictionary(dict) {
		if (!sf.ActionInterval.prototype.initWithDictionary.call(this, dict)) {
			return false;
		}
		var action = sf.Action.subActionFromDictionary(dict);
		if (action instanceof cc.FiniteTimeAction) {
			this._action = action;
			return true;
		}
		
		cn.error(dict);
		return false;
	}
	
	function ccAction() {
		if (!this._action) {
			cn.error(this._action);
			return null;
		}
		return cc.ReverseTime.create(this._action);
	}
	
	var prop = {
		ctor: ReverseTime,
		
		init: init,
		initWithDictionary: initWithDictionary,
		
		ccAction: ccAction,
		
		// properties
		_action: null,
	};
	
	// extends
	sf.ReverseTime = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

//====================================================================== Animate

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function Animate() {
		sf.ActionInterval.prototype.ctor.call(this);
		this._frames = null;
		this._delay = 0;
		this._loops = 1;
		this._restoreOriginalFrame = false;
	}
	
	//
	//  static functions
	//
	
	//
	//  instance functions
	//
	function init() {
		if (!sf.ActionInterval.prototype.init.call(this)) {
			return false;
		}
		this._frames = null;
		this._delay = 0;
		this._loops = 1;
		this._restoreOriginalFrame = false;
		return true;
	}
	function initWithDictionary(dict) {
		if (!sf.ActionInterval.prototype.initWithDictionary.call(this, dict)) {
			return false;
		}
		
		var frames = dict['Frames'];
		this._frames = frames ? frames : [];
		
		var delay = dict['Delay'];
		this._delay = delay;
		
		var loops = dict['Loops'];
		this._loops = loops !== undefined ? loops : 1;
		
		var restoreOriginalFrame = dict['RestoreOriginalFrame'];
		this._restoreOriginalFrame = restoreOriginalFrame;
		
		return true;
	}
	
	function createAnimation(frames, delayTime) {
		
		var spriteFrames = [];
		var frame, file;
		for (var i = 0; i < frames.length; ++i) {
			file = frames[i];
			if (!file) {
				continue;
			} else if (typeof file === 'string') {
				// frame name
			} else {
				// file name
				file = file['File'];
			}
			if (!file) {
				continue;
			}
			
			frame = cn.Sprite.getSpriteFrameWithFile(file);
			if (frame) {
				spriteFrames.push(frame);
			} else {
				cn.error(file, frame, frames[i]);
			}
		}
		
		return cc.Animation.create(spriteFrames, delayTime);
	}
	
	function ccAction() {
		var frames = this._frames;
		if (!frames || frames.length < 1) {
			cn.error(this._frames);
			return null;
		}
		var delayTime = this._duration ? this._duration / frames.length : this._delay;
		var animation = createAnimation(frames, delayTime);
		if (!animation) {
			cn.error(frames, delayTime, this._loops);
			return null;
		}
		animation.setLoops(this._loops);
		animation.setRestoreOriginalFrame(this._restoreOriginalFrame);
		return cc.Animate.create(animation);
	}
	
	var prop = {
		ctor: Animate,
		
		init: init,
		initWithDictionary: initWithDictionary,
		
		ccAction: ccAction,
		
		// properties
		_frames: null,
		_delay: null,
		_loops: null,
		_restoreOriginalFrame: null,
	};
	
	// extends
	sf.Animate = sf.ActionInterval.extend(prop);
	
}(SpriteForest);

File: objc/action/SFActionInterval.m
//
//  SFActionInterval.m
//  SpriteForest2
//
//  Created by Moky on 12-10-1.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFActionInterval.h"

#import "coconut2d.h"
#import "CCMask.h"
#import "SFAnimation.h"


@implementation SFActionInterval

- (CCAction *) ccAction
{
	CNLog(@"override me");
	return nil;
//	return [CCActionInterval actionWithDuration:duration_];
}

@end


@implementation SFSequence

@synthesize actions = actions_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		NSArray * actions = [[self class] subActionsFromDictionary:dict];
		if ([actions isKindOfClass:[NSArray class]])
		{
			NSMutableArray * mArray = [NSMutableArray arrayWithCapacity:[actions count]];
			for (id action in actions)
			{
				if ([action isKindOfClass:[CCFiniteTimeAction class]])
				{
					[mArray addObject:action];
				}
			}
			if ([mArray count] > 0)
			{
				self.actions = mArray;
			}
		}// EOF if ([actions isKindOfClass:[NSArray class]])
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.actions = nil;
	}
	return self;
}

- (void) dealloc
{
	[actions_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![actions_ isKindOfClass:[NSArray class]])
	{
		CNLog(@"Error: actions = %@", actions_);
		return [super ccAction];
	}
	return [CCSequence actionsWithArray:actions_];
}

@end


@implementation SFRepeat

@synthesize action = action_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		// drived action
		id action = [[self class] subActionFromDictionary:dict];
		if ([action isKindOfClass:[CCFiniteTimeAction class]])
		{
			self.action = action;
		}
		
		// times
		times_ = [[dict objectForKey:@"Times"] intValue];
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.action = nil;
	}
	return self;
}

- (void) dealloc
{
	[action_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCFiniteTimeAction class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCRepeat actionWithAction:action_ times:times_];
}

@end


@implementation SFSpawn

@synthesize actions = actions_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		NSArray * actions = [[self class] subActionsFromDictionary:dict];
		if ([actions isKindOfClass:[NSArray class]])
		{
			NSMutableArray * mArray = [NSMutableArray arrayWithCapacity:[actions count]];
			for (id action in actions)
			{
				if ([action isKindOfClass:[CCFiniteTimeAction class]])
				{
					[mArray addObject:action];
				}
			}
			if ([mArray count] > 0)
			{
				self.actions = mArray;
			}
		}// EOF if ([actions isKindOfClass:[NSArray class]])
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.actions = nil;
	}
	return self;
}

- (void) dealloc
{
	[actions_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![actions_ isKindOfClass:[NSArray class]])
	{
		CNLog(@"Error: actions = %@", actions_);
		return [super ccAction];
	}
	return [CCSpawn actionsWithArray:actions_];
}

@end


@implementation SFRotateTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id val = [dict objectForKey:@"Angle"] ? [dict objectForKey:@"Angle"] : [dict objectForKey:@"Rotation"];
		angle_ = [val floatValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCRotateTo actionWithDuration:duration_ angle:angle_];
}

@end


@implementation SFRotateBy

- (CCAction *) ccAction
{
	return [CCRotateBy actionWithDuration:duration_ angle:angle_];
}

@end


@implementation SFMoveTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		position_ = [dict objectForKey:@"Position"] ? CCPointFromString([dict objectForKey:@"Position"]) : CGPointZero;
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCMoveTo actionWithDuration:duration_ position:position_];
}

@end


@implementation SFMoveBy

- (CCAction *) ccAction
{
	return [CCMoveBy actionWithDuration:duration_ position:position_];
}

@end


@implementation SFSkewTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		skewX_ = [[dict objectForKey:@"SkewX"] floatValue];
		skewY_ = [[dict objectForKey:@"SkewY"] floatValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCSkewTo actionWithDuration:duration_ skewX:skewX_ skewY:skewY_];
}

@end


@implementation SFSkewBy

- (CCAction *) ccAction
{
	return [CCSkewBy actionWithDuration:duration_ skewX:skewX_ skewY:skewY_];
}

@end


@implementation SFJumpTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		position_ = [dict objectForKey:@"Position"] ? CCPointFromString([dict objectForKey:@"Position"]) : CGPointZero;
		height_ = [[dict objectForKey:@"Height"] floatValue];
		jumps_ = [[dict objectForKey:@"Jumps"] intValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCJumpTo actionWithDuration:duration_ position:position_ height:height_ jumps:jumps_];
}

@end


@implementation SFJumpBy

- (CCAction *) ccAction
{
	return [CCJumpBy actionWithDuration:duration_ position:position_ height:height_ jumps:jumps_];
}

@end


@implementation SFBezierTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		bezierConfig_.endPosition = [dict objectForKey:@"EndPosition"] ? CCPointFromString([dict objectForKey:@"EndPosition"]) : CGPointZero;
		bezierConfig_.controlPoint_1 = [dict objectForKey:@"ControlPoint1"] ? CCPointFromString([dict objectForKey:@"ControlPoint1"]) : CGPointZero;
		bezierConfig_.controlPoint_2 = [dict objectForKey:@"ControlPoint2"] ? CCPointFromString([dict objectForKey:@"ControlPoint2"]) : CGPointZero;
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCBezierTo actionWithDuration:duration_ bezier:bezierConfig_];
}

@end


@implementation SFBezierBy

- (CCAction *) ccAction
{
	return [CCBezierBy actionWithDuration:duration_ bezier:bezierConfig_];
}

@end


@implementation SFScaleTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id scale = [dict objectForKey:@"Scale"];
		if (scale)
		{
			scaleX_ = [scale floatValue];
			scaleY_ = [scale floatValue];
		}
		else
		{
			scaleX_ = [[dict objectForKey:@"ScaleX"] floatValue];
			scaleY_ = [[dict objectForKey:@"ScaleY"] floatValue];
		}
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCScaleTo actionWithDuration:duration_ scaleX:scaleX_ scaleY:scaleY_];
}

@end


@implementation SFScaleBy

- (CCAction *) ccAction
{
	return [CCScaleBy actionWithDuration:duration_ scaleX:scaleX_ scaleY:scaleY_];
}

@end


@implementation SFBlink

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		times_ = [[dict objectForKey:@"Times"] intValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCBlink actionWithDuration:duration_ blinks:times_];
}

@end


@implementation SFFadeIn

- (CCAction *) ccAction
{
	return [CCFadeIn actionWithDuration:duration_];
}

@end


@implementation SFFadeOut

- (CCAction *) ccAction
{
	return [CCFadeOut actionWithDuration:duration_];
}

@end


@implementation SFFadeTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		opacity_ = [[dict objectForKey:@"Opacity"] intValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCFadeTo actionWithDuration:duration_ opacity:opacity_];
}

@end


@implementation SFTintTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		red_ = [[dict objectForKey:@"Red"] intValue];
		green_ = [[dict objectForKey:@"Green"] intValue];
		blue_ = [[dict objectForKey:@"Blue"] intValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCTintTo actionWithDuration:duration_ red:red_ green:green_ blue:blue_];
}

@end


@implementation SFTintBy

- (CCAction *) ccAction
{
	return [CCTintBy actionWithDuration:duration_ red:red_ green:green_ blue:blue_];
}

@end


@implementation SFDelayTime

- (CCAction *) ccAction
{
	return [CCDelayTime actionWithDuration:duration_];
}

@end


@implementation SFReverseTime

@synthesize action = action_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id action = [[self class] subActionFromDictionary:dict];
		if ([action isKindOfClass:[CCActionInterval class]])
		{
			self.action = action;
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.action = nil;
	}
	return self;
}

- (void) dealloc
{
	[action_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCFiniteTimeAction class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCReverseTime actionWithAction:action_];
}

@end


@implementation SFMaskTo

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		rect_ = [dict objectForKey:@"Rect"] ? CCRectFromString([dict objectForKey:@"Rect"]) : CGRectZero;
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCMaskTo actionWithDuration:duration_ rect:rect_];
}

@end


@implementation SFMaskBy

- (CCAction *) ccAction
{
	return [CCMaskBy actionWithDuration:duration_ rect:rect_];
}

@end


@implementation SFAnimate

@synthesize frames = frames_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id frames = [dict objectForKey:@"Frames"];
		if ([frames isKindOfClass:[NSArray class]])
		{
			self.frames = frames;
		}
		delay_ = [[dict objectForKey:@"Delay"] floatValue];
		restoreOriginalFrame_ = [[dict objectForKey:@"RestoreOriginalFrame"] boolValue];
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.frames = nil;
	}
	return self;
}

- (void) dealloc
{
	[frames_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	CGFloat delaytime = (duration_ * [frames_ count]) ? duration_ / [frames_ count] : delay_;
	CCAnimation * animation = [SFAnimation animationWithFrames:frames_ delay:delaytime];
	if (![animation isKindOfClass:[CCAnimation class]])
	{
		CNLog(@"Error: frames = %@, delay = %.2f, restoreOriginalFrame = %d", frames_, delay_, restoreOriginalFrame_);
		return [super ccAction];
	}
	return [CCAnimate actionWithDuration:duration_ animation:animation restoreOriginalFrame:restoreOriginalFrame_];
}

@end
